### PR TITLE
Vaccination by age

### PR DESCRIPTION
**Pull request by annecori**
This builds on (and modifies) #139.

- Relative susceptibility due to vaccine can now be age dependant

- Vaccination and building up/waning of immunity now implemented, i.e. individuals can move between different S compartments, with parameters which can be age dependant. 

- Parameters governing movement between S compartments are not time varying at the moment

- For this to work I had to force having at least 3 vaccination classes, but by default the second and third and flows to those are empty / zero. NOTE THIS IS LIKELY TO BREAK THE ORDERLY TASKS

Would be great to discuss and if we want to make this part of master, merge asap before master diverges again.

### CODE REVIEW

**Review by richfitz — CHANGES_REQUESTED** [review]
I'm happy with this from a code perspective, with minor comments included here. @edknock and others should comment about how this affects the model, and I believe that this does cause some issues with the seroreversion PR #143

File: R/carehomes.R [hunk 1]
```diff
@@ -48,10 +48,28 @@ NULL
...
+##'   vaccination group (columns of the matrix); in that case, in each row of
+##'   the matrix, the first value should be 1 (for the non-vaccinated group)
+##'   and subsequent values be between 0 and 1
+##'
+##' @param vaccination_rate A single value or vector of values representing the
+##'   rate of vaccination of susceptibles. If a single value, the same rate is
+##'   used for all age groups; if a vector, its length needs to match the number
+##'   of age groups and it should contain age-specific vaccination rates
+##'
+##' @param vaccine_progression_rate A vector or matrix of values representing
```
[the remark this answers is not in the document]
Comment (annecori) [reply R/carehomes.R:67]: Good point. I've added some [examples](https://github.com/mrc-ide/sircovid/blob/84b5dbf4f7836fbbaac0c70c66b2e39b62a4258a/man/carehomes_parameters.Rd#L130) - do you think this is enough?

**Comment by annecori** [comment]
I have addressed all of Rich's comments on this PR and have created a sister PR [#841](https://github.com/ncov-ic/ncov-outputs/pull/841) in ncos-outputs. @edknock and/or @lwhittles when you have a chance to look at those two it would be great. Thx

File: R/carehomes.R [hunk 2]
```diff
@@ -66,6 +84,75 @@ NULL
...
+##' # vaccine-induced immunity wanes after a period which is exponentially
+##' # distributed and lasts on average 26 weeks (half a year);
+##' # there are similar across all age groups
+##' vaccine_progression_rate <- c(1/(2*7), 1/(26*7))
+##'
+##' # generate model parameters
+##' p <- carehomes_parameters(sircovid_date("2020-02-01"), "uk",
+##' rel_susceptibility = rel_susceptibility,
+##' vaccination_rate = vaccination_rate,
+##' vaccine_progression_rate = vaccine_progression_rate)
```
Comment (richfitz) [inline R/carehomes.R:116]:
```suggestion
##' p <- carehomes_parameters(sircovid_date("2020-02-01"), "uk",
##'                           rel_susceptibility = rel_susceptibility,
##'                           vaccination_rate = vaccination_rate,
##'                           vaccine_progression_rate = vaccine_progression_rate)
```

File: DESCRIPTION [hunk 3]
```diff
@@ -1,6 +1,6 @@
 Package: sircovid
 Title: SIR Model for COVID-19
-Version: 0.6.6
+Version: 0.6.7
```
Comment (richfitz) [inline DESCRIPTION:3]:
```suggestion
Version: 0.6.8
```

Can you also add a line in NEWS.md (see the format there)

File: DESCRIPTION [hunk 4]
```diff
@@ -29,7 +29,11 @@ Authors@R:
       person(given = "OJ",
            family = "Watson",
            role = c("ctb"),
-           email = "o.watson15@imperial.ac.uk"))
+           email = "o.watson15@imperial.ac.uk"),
+      person(given = "Anne",
+           family = "Cori",
+           role = c("ctb"),
```
Comment (richfitz) [inline DESCRIPTION:35]:
```suggestion
           role = c("aut"),
```

let's be real here

**Review by richfitz — CHANGES_REQUESTED** [review]
Small things here. Go to the "files" tab and you can batch these into a single commit (or do them one-by-one). Then redocument. If Ed is happy from the science side we can merge this first thing Monday

File: inst/odin/carehomes.R [hunk 5]
```diff
@@ -113,7 +120,12 @@ prob_admit_conf[] <- p_admit_conf * psi_admit_conf[i]
 
 ## Draws from binomial distributions for numbers changing between
 ## compartments:
-n_SE[, ] <- rbinom(S[i, j], p_SE[i, j])
+
+# competing risk of infection and progression through vaccine classes
+n_S_out[, 1:n_vacc_classes_minus_1] <- rbinom(S[i, j], p_SE[i, j] + p_SS[i, j])
```
Comment (edknock) [inline inst/odin/carehomes.R:125]: Surely the SE and SS transitions are not mutually exclusive? Unless I'm mistaken, this calculation doesn't seem to account for the intersection of these transitions.

**Review by edknock — CHANGES_REQUESTED** [review]

File: inst/odin/carehomes.R (hunk 5 above)
Comment (annecori) [reply inst/odin/carehomes.R:125]: agreed - I think for a small enough timestep it is ok to assume no-one gets vaccinated and infected in the same time step (or changes vaccination category and gets infected in the same timestep). @edknock and @MJomaba thoughts about this?

File: tests/testthat/test-carehomes-vaccination.R [hunk 6]
```diff
@@ -31,8 +31,61 @@ test_that("there are no infections if everyone is vaccinated with a vaccine
...
+                                      vaccine_progression_rate = rep(1e9, 10))
+            mod <- carehomes$new(p, 0, 1)
+            info <- mod$info()
+            mod$set_state(carehomes_initial(info, 1, p)$state)
+            mod$set_index(integer(0))
+            y <- mod$transform_variables(drop(
+              dust::dust_iterate(mod, seq(0, 400, by = 12))))
+            expect_true(all(y$S[, 1, 1] == y$S[, 12, 2]))
+            expect_true(all(y$S[, , 34] == y$S[, , 2]))
+})
```
[the remark this answers is not in the document]
Comment (annecori) [reply tests/testthat/test-carehomes-vaccination.R:86]: good idea - and this test will need to be updated when we model the impact of vaccination on E, I and R as well as S. 
I've added a test [here](https://github.com/mrc-ide/sircovid/blob/060fa316101cbddc0492d417167e82777f42e234/tests/testthat/test-carehomes-vaccination.R#L88).

File: inst/odin/carehomes.R (hunk 5 above)
Comment (edknock) [reply inst/odin/carehomes.R]: Happy now with how this has been revised.

**Comment by edknock** [comment]
Happy with the changes made since my previous review

**Review by edknock — APPROVED** [review]

**Review by richfitz — APPROVED** [review]

**Pull request merged**